examples/Superresolution/enet-pat.py

```diff
@@ -0,0 +1,330 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Author: Patrick Wieschollek <mail@patwie.com>
+
+import os
+import argparse
+import cv2
+import six
+import numpy as np
+import tensorflow as tf
+from tensorpack import *
+from tensorpack.tfutils.scope_utils import auto_reuse_variable_scope
+from GAN import MultiGPUGANTrainer, GANModelDesc
+from tensorpack.tfutils.summary import add_moving_summary
+from tensorpack.utils import logger
+from data_sampler import ImageDecode
+
+
+"""
+Re-implementation:
+EnhanceNet: Single Image Super-Resolution Through Automated Texture Synthesis
+Sajjadi et al. <https://arxiv.org/abs/1612.07919>, ICCV 2017
+
+
+train:
+
+    wget http://images.cocodataset.org/zips/train2017.zip
+    python data_sampler.py --lmdb train2017.lmdb --input train2017.zip --create
+    python enet-pat.py --vgg19 /path/to/vgg19.npy --gpu 0,1 --lmdb train2017.lmdb
+
+eval:
+
+    python enet-pat.py --apply --load /checkpoints/checkpoint --lowres "eagle.png" --output "eagle" --gpu 1
+"""
+
+BATCH_SIZE = 6
+CHANNELS = 3
+SHAPE_LR = 32
+NF = 64
+VGG_MEAN = np.array([123.68, 116.779, 103.939])  # RGB
+
+
+def normalize(v):
+    assert isinstance(v, tf.Tensor)
+    v.get_shape().assert_has_rank(4)
+    dim = v.get_shape().as_list()
+    return v / (dim[1] * dim[2] * dim[3])
+
+
+def gram_matrix(v):
+    assert isinstance(v, tf.Tensor)
+    v.get_shape().assert_has_rank(4)
+    dim = v.get_shape().as_list()
+    v = normalize(v)
+    v = tf.reshape(v, [dim[0], dim[1] * dim[2], dim[3]])
+    return tf.matmul(v, v, transpose_a=True)
+
+
+class Model(GANModelDesc):
+
+    def __init__(self, height=SHAPE_LR, width=SHAPE_LR, inference=False):
+        super(Model, self).__init__()
+        self.height = height
+        self.width = width
+        self.inference = inference
```

Comment:
why this? seems like `inference=ctx.is_training` is always true

Reply:
not during "apply". There is a difference subtle difference between a callback inference runner and the `apply` function. The inference runner might need the loss statistics, the `appply` function not.

Reply:
OfflinePredictor always creates model with ctx.is_training=False

Reply:
InferenceRunnerCallback?

Reply:
I understand the need, but you're not using inferencerunner

Reply:
fixed (as I do not use the InferenceRunner here). Maybe use `ctx.phase = ['train', 'val', 'inference']` at some point

Reply:
Well yeah, I also felt having these two phases isn't perfect, though users can always do their own stuff like you did.

Reply:
I had this issue not before. As long as GANs are not stable (requiring the InferenceRunner), we do not need these different phases :-)